src/schema/v2/fields/hybridConnection/hybridOffsets.ts

```diff
@@ -0,0 +1,85 @@
+import { URLSearchParams } from "url"
+import { ConnectionCursor } from "graphql-relay"
+import qs from "qs"
+import { unBase64, base64 } from "lib/base64"
+
+type OffsetState<T extends string> = Record<T, number> & {
+  // this is kind of strange; the _position attr is the overall
+  // (0-based) index in the collection. It starts out null
+  _position: number | null
+}
+
+const PREFIX = "offsets"
+
+// increment a number input or `null` to 0
+const incrementFromNull = (x: number | null): number => (x === null ? 0 : x + 1)
+
+export class HybridOffsets<T extends string> {
+  readonly state: OffsetState<T>
+
+  /**
+   * create an empty HybridOffsets with the supplied source keys plus `position`.
+   * Each offset key is initialized to 0, while `position` is a 0-indexed counter
+   * of the entire collection and thus starts at `null`, then 0, 1, etc.
+   */
+  static empty<U extends string>(keys: Array<U>): HybridOffsets<U> {
+    const offsets = keys.reduce<OffsetState<U>>(
+      (acc: OffsetState<U>, key: U) => {
+        return { ...acc, [key]: 0 }
+      },
+      { _position: null } as OffsetState<U>
+    )
+    return new HybridOffsets(offsets)
```

Comment:
Here, as elsewhere, its a good habit to assign to a variable and return variable versus returning an operation